daemon/emer-daemon.c

```diff
@@ -471,21 +473,21 @@ queue_http_request (GTask *upload_task)
       return;
     }
 
-  SoupURI *http_request_url =
+  g_autoptr(GUri) http_request_url =
     get_http_request_url (self, serialized_request_body,
                           serialized_request_body_length);
-  SoupMessage *http_message =
+  g_autoptr(SoupMessage) http_message =
     soup_message_new_from_uri ("PUT", http_request_url);
-  soup_uri_free (http_request_url);
 
-  soup_message_headers_append (http_message->request_headers,
+  soup_message_headers_append (soup_message_get_request_headers (http_message),
                                "X-Endless-Content-Encoding", "gzip");
-  soup_message_set_request (http_message, "application/octet-stream",
-                            SOUP_MEMORY_TAKE, compressed_request_body,
-                            compressed_request_body_length);
-  soup_session_queue_message (self->http_session, http_message,
-                              (SoupSessionCallback) handle_http_response,
-                              upload_task);
+
+  g_autoptr(GBytes) request_body = g_bytes_new (compressed_request_body, compressed_request_body_length);
+  soup_message_set_request_body_from_bytes (http_message, "application/octet-stream", request_body);
+
+  soup_session_send_async (self->http_session, http_message, G_PRIORITY_DEFAULT, NULL,
```

Comment:
I'm on the fence about this. On the one hand, I don't think we necessarily want to be able to cancel these messages. On the other hand, `handle_http_response` doesn't handle the case where the request comes after `EmerDaemon` is destroyed... I'd prefer making `handle_http_response` handle destructed objects, but it's probably tricky, so cancelling on shutdown might be easier. Thoughts?

Reply:
I was on the fence as well :) At first, I chained it up with the GTask's cancellable, but that ended up breaking self tests by (if I remember it right) reversing the order in which cancelled messages were handled or something like that. I gave up and just took it out at that point, thinking that it would make it easier to do libsoup 2 ifdefs if I don't rework the code too much.

I can look at it again if you think it makes sense, or should we wait until libsoup 2 ifdefs are gone? It's my first libsoup 3 port so I'm unsure how to handle things in a clean way and keep libsoup 2 compatibility at the same time.

Reply:
The `EmerDaemon` can't be destroyed until after `handle_http_response` is called, because the `upload_task` holds a strong reference to it. I agree that it would be better to chain the task's cancellable through but it's not a regression not to. IMO this is OK for now.